pulsar-metadata/src/main/java/org/apache/pulsar/metadata/impl/RocksdbMetadataStore.java

```diff
@@ -0,0 +1,428 @@
+
+    @Data
+    @AllArgsConstructor
+    @NoArgsConstructor
+    private static class MetaValue {
+        private static final int HEADER_SIZE = 8 + 8 + 8 + 8 + 1;
+
+        long version;
+        long owner;
+        long createdTimestamp;
+        long modifiedTimestamp;
+        boolean ephemeral;
+        byte[] data;
+
+        public byte[] serialize() {
+            byte[] result = new byte[HEADER_SIZE + data.length];
+            ByteBuffer buffer = ByteBuffer.wrap(result);
+            buffer.putLong(version);
+            buffer.putLong(owner);
+            buffer.putLong(createdTimestamp);
+            buffer.putLong(modifiedTimestamp);
+            buffer.put((byte) (ephemeral ? 1 : 0));
+            buffer.put(data);
+            return result;
+        }
+
+        public static MetaValue parse(byte[] dataBytes) throws MetadataStoreException {
+            if (dataBytes == null) {
+                return null;
+            }
+            if (dataBytes.length < HEADER_SIZE) {
+                throw new MetadataStoreException("Invalid MetaValue data");
+            }
+            ByteBuffer buffer = ByteBuffer.wrap(dataBytes);
+            MetaValue metaValue = new MetaValue();
+            metaValue.version = buffer.getLong();
+            metaValue.owner = buffer.getLong();
+            metaValue.createdTimestamp = buffer.getLong();
+            metaValue.modifiedTimestamp = buffer.getLong();
+            metaValue.ephemeral = buffer.get() > 0;
+            metaValue.data = new byte[buffer.remaining()];
+            buffer.get(metaValue.data);
+            return metaValue;
+        }
+    }
+
+    @VisibleForTesting
+    static byte[] toBytes(String s) {
+        return s.getBytes(StandardCharsets.UTF_8);
+    }
+
+    @VisibleForTesting
+    static String toString(byte[] bytes) {
+        return new String(bytes, StandardCharsets.UTF_8);
+    }
+
+    @VisibleForTesting
+    static byte[] toBytes(long value) {
+        return ByteBuffer.wrap(new byte[8]).putLong(value).array();
+    }
+
+    @VisibleForTesting
+    static long toLong(byte[] bytes) {
+        return ByteBuffer.wrap(bytes).getLong();
+    }
+
+    /**
+     * @param metadataURL         format "rocksdb://{storePath}"
+     * @param metadataStoreConfig
+     * @throws MetadataStoreException
+     */
+    public RocksdbMetadataStore(String metadataURL, MetadataStoreConfig metadataStoreConfig)
+            throws MetadataStoreException {
+        try {
+            RocksDB.loadLibrary();
+        } catch (Throwable t) {
+            throw new MetadataStoreException("Failed to load RocksDB JNI library", t);
+        }
+
+        this.optionSync = new WriteOptions();
+        this.optionDontSync = new WriteOptions();
+        this.optionCache = new ReadOptions();
+        this.optionDontCache = new ReadOptions();
+        this.emptyBatch = new WriteBatch();
+
+        try (Options options = new Options()) {
+            options.setCreateIfMissing(true);
```

Comment:
We should take a Rocksdb config file path argument from our config, so that a user will be able to control all the aspects of RocksDB.

Reply:
Add a new config filed `metadataStoreConfigPath` in ServiceConfiguration.